docs/src/internals.md

```diff
@@ -0,0 +1,256 @@
+log(a+b))` is converted into the anonymous function `(a,b) -> log(a+b)`.
+
+### Schema time
+
+The next phase of life for a formula begins when a _schema_ for the data becomes
+available.  A schema is a mapping from data columns to a concrete term
+type---either a `ContinuousTerm` or a `CategoricalTerm`---which represents all
+the summary information about a data column necessary to create a model matrix
+from that column.
+
+A schema is computed with the `schema` function.  By default, it will create a
+schema for every column in the data:
+
+```@repl 1
+using DataFrames    # for pretty printing---any Table will do
+df = DataFrame(y = rand(9), a = [1:9;], b = rand(9), c = repeat(["a","b","c"], 3))
+schema(df)
+```
+
+However, if a term (including a `FormulaTerm`) is provided, the schema will be
+computed based only on the necessary variables:
+
+```@repl 1
+schema(@formula(y ~ 1 + a), df)
+schema(Term(:a) + Term(:b), df)
+```
+
+Once a schema is computed, it's _applied_ to the formula with
+[`apply_schema`](@ref).  Among other things, this _instantiates_ placeholder
+terms: 
+* `Term`s become `ContinuousTerm`s or `CategoricalTerm`s
+* `ConstantTerm`s become `InterceptTerm`s
+* Tuples of terms [`MatrixTerm`](@ref)s where appropriate to explicitly indicate
+  they should be concatenated into a single model matrix
+
+```@repl 1
+f = @formula(y ~ 1 + a + b + c)
+typeof(f)
+f = apply_schema(f, schema(f, df))
+typeof(f)
+```
+
+This transformation is done by calling `apply_schema(term, schema, ModelType)`
+recursively on each term (`ModelType` defaults to `StatisticalModel`).  Because
+`apply_schema` dispatches on the term, schema, and model type, this stage allows
+generic context-aware transformations, based on _both_ the source (schema) _and_
+the destination (model type).  This is the primary mechanisms by which the
+formula DSL can be extended ([see below](#Extending-@formula-syntax-1) for more
+details)
+
+### Data time
+
+At the end of "schema time", a formula encapsulates all the information needed
+to convert a table into a numeric model matrix.  That is, it is ready for "data
+time".  The main API method is [`model_cols`](@ref), which when applied to a
+`FormulaTerm` returns a tuple of the numeric forms for the left- (response) and
+right-hand (predictor) sides.
+
+```@repl 1
+model_cols(f, df)
+```
+
+Model cols can also take a single row from a table, as a `NamedTuple`:
+
+```@repl 1
+using Tables
+model_cols(f, first(Tables.rowtable(df)))
+```
+
+Any `AbstractTerm` can be passed to `model_cols` with a table and returns one or
+more numeric arrays:
+
+
+
+## Extending `@formula` syntax
+
+Package authors may want to create additional syntax to the `@formula` DSL so
+their users can conveniently specify particular kinds of models.  StatsModels.jl
+provides mechanisms for such extensions that do _not_ rely on compile time
+"macro magic", but on standard julian mechanisms of multiple dispatch.
+
+Extensions have three components:
```

Comment:
According to the summary below, each of these applies to one of the formula steps. It would be useful to mention this.